### ISSUE TITLE
Behavior of --except argument is frustrating and confusing

### ISSUE THREAD

**Issue by cs-shadow** [cs-shadow/buildstream#135]
In GitLab by @tristanvb on Oct 26, 2017, 08:17

Currently, given any recursive operation with an `--except` option, the behavior is to scan the loaded pipeline and remove the except dependencies recursively.

The problem is once you have a fairly large project, like the GNOME modulesets for instance, there are many elements to choose from, when I run a command such as:
```
bst track --deps all --except core/meta-gnome-core.bst apps/meta-gnome-apps-tested.bst
```

I naturally expect that BuildStream will track all of the `meta-gnome-apps-tested.bst` except for the elements in `core/meta-gnome-core.bst`, these two virtual targets (stacks), basically represent what we release in terms of the apps layer and the core layer, *however*, `apps/meta-gnome-apps-tested.bst` does not depend on `core/meta-gnome-core.bst`.

This is impractical, any larger projects will run into this, and it's too tedious to examine the common dependencies of both virtual targets and supply `--except` options for each intersection manually, this defeats the convenience purpose of the `--except` option altogether.

To make matters worse, the error message the user receives in this case is `No element named core/meta-gnome-core.bst`, leading the user to suspect a buildstream bug, since clearly there *is* an element with that name in their project.

Instead; BuildStream should actually load the `--except` targets specified, and subtract the intersection from the pipeline targets - making the above example invocation *just work*.

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Oct 26, 2017, 08:24

mentioned in commit 640e7e57842ca4c2ec3a8641258f54ff7c42d3c0

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Oct 26, 2017, 08:25

To slightly alleviate the pain, I've adjusted the error message to say:

> `No element named core/meta-gnome-core.bst in the loaded pipeline`

**Comment by cs-shadow** [comment]
In GitLab by @tlater on Nov 10, 2017, 10:51

created branch [`131-behavior-of-except-argument-is-frustrating-and-confusing`](https://gitlab.com/BuildStream/buildstream/compare/master...131-behavior-of-except-argument-is-frustrating-and-confusing)

**Comment by cs-shadow** [comment]
In GitLab by @tlater on Nov 10, 2017, 10:51

mentioned in merge request !147

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Nov 18, 2017, 09:58

closed via commit 8e3bcb5517896480802360abf22d6d464c0b5179

**Comment by cs-shadow** [comment]
In GitLab by @tristanvb on Nov 18, 2017, 17:27

mentioned in commit 8e3bcb5517896480802360abf22d6d464c0b5179
